all: Support multiple device connections (fixes #141)

calmh opened the pull request:
### Purpose

This adds the ability to have multiple concurrent connections to a single device. This is primarily useful when the network has multiple physical links for aggregated bandwidth. A single connection will never see a higher rate than a single link can give, but multiple connections are load-balanced over multiple links.

It is also incidentally useful for older multi-core CPUs, where bandwidth could be limited by the TLS performance of a single CPU core -- using multiple connections achieves concurrency in the required crypto calculations...

In principle, it could also be helpful in situations with high latency or random packet loss, but I haven't tested that.

### Testing

I've been running it myself for a while. It's also being run/tested in a separate, larger (actual) setup.

### Screenshots

<img width="576" alt="Screenshot 2023-06-01 at 10 07 41" src="https://github.com/syncthing/syncthing/assets/125426/f31fc13e-970e-4042-82d4-355facf9dfdc">


### Documentation

See:

https://github.com/syncthing/docs/pull/803/files

(seriously, take a peek, it describes some things that are useful in understanding the PR)

TODO: Describe the new extended /rest/system/connections output

### Implementation notes

- I added a new field `bool secondary` to the cluster config, and secondary connections get a cluster config with just this field set and nothing else. This declares the intention for the connection when it starts, and fulfils the existing requirement that the first message must be a cluster config.
- Protocol connections gain a new `ConnectionID()` which is a unique identifier per connection so that we can refer to them where we previously used just "the connection to $deviceID".
- Further refactoring in the protocol package:
	- The `(protocol.Connection).ID()` that returned the device ID is now called `DeviceID()` to disambiguate from `ConnectionID()`.
	- The callbacks that used to get a device ID now get the actual connection instead, which I think makes more sense and gives the callee more power. However, the actual protocol methods that call the callbacks don't know about the Connection (they just operate on a read-writer) so I added a separate wrapper outside to populate the connection parameter, and the internal callbacks are simpler.
	- The above required a lot of simple changes to the tests.
- The model now keeps track of all active connections, and "promotes" one of the connections to being the "primary": where we send indexes and cluster configs. Which connection this is can change over time.
- This promotion happens with a slight delay, because it's better we don't run it eight times in quick succession when a device with eight connections restart, for example.
- Some things that previously happened when adding a connection now happen slightly delayed in the promotion loop instead. Adding (or removing a connection) schedules the promotion loop.
- Index handling in the model is slightly changed, we now create an index handler when we get a cluster config instead of when we add a connection.

### Also

Fixes #4202
Fixes #4207
Fixes #8184
Fixes #8731

bt90 commented:
Very nice. This should fix #4207 and #8184

What's the default number of additional connections?

Could we be a bit more clear in the UI though?

Maybe something like `1 control + 7 data` ?

calmh commented:
The default is one connection, as it's always been. With that config there is nothing extra shown in the GUI either.

bt90 commented:
What's the downside of enabling this by default with a reasonable number of connections? The topic came up quite often in the past and I fear that we're heading in the direction of nginx which suffers from bad defaults and configuration pitfalls. Ideally a user should only need to change settings for uncommon scenarios.

ISPs applying throttling per connection is sadly quite common and from a support point of view, i'd rather not have to post the same documentation link over and over again **if** this feature is fine for the majority of the userbase.

calmh commented:
There should be no need under normal circumstances, and in my opinion ISP throttling is a really bad use case for this. Any decent ISP will apply shaping per-customer or at least per-IP which this will not help against, and if does it fails the "what if everyone did this" check. In fact that use case has been the number one thing to make me hesitate to do this, but now there have surfaced other use cases that are more compelling to me...

imsodin commented:
I had a first look, not yet close to actually having checked out and grasping all of it.

Basically my first thought upen glancing at the model was: Does the model really need to handle this directly?  
I am nowhere close to understanding it well enough to see if it's necessary or to have a concrete idea of how to separate it - just mentioning it for now. I'll see how the thought develops when I come back to have a better look at how it works in the model.

Another purely procedural thing: Some of the renaming/refactoring causes a lot of noise in the diff, is also quite independent and sometimes would even make sense without the multiple connection things (e.g. `ID` -> `DeviceID`). Would be nice/help reviewing if those were separate commits (and some maybe even PRs to get them merged/out of the way entirely).

calmh commented:
> Basically my first thought upen glancing at the model was: Does the model really need to handle this directly?

Mybae, maybe not, let me explain the reasons why I did what I did. I had the same thought as you, probably, and initially tries/wanted to create something that looked like one connection to the model etc but did multiplexing under the hood. This gets complicated though because, to begin with, we can't establish new connections from the protocol code, it doesn't know how (it's just a ReadWriter at that point) so this still needs to happen in the connections service, which is also the only place we have the socker for dialling etc. It then needs to track all connections and somehow add a new sub-connection to some sort of primary connection.

This is all doable, of course, but we also need to take some care when connections drop, because if we sent an IndexUpdate and then that connection gets reset we don't know if it reached the recipient or not. If we're trying to look like a single connection, abstraction wise, the model doesn't get to know that a connection dropped and can't advise us on the current state of index handlers etc. Same thing with requests and responses, really, just with less disastrous effects -- we'd need to implement some sort of extra reliability layer on top of potentially unreliable individual connections, and, eh, I'd rather not.

With the implementation here all of that is simpler, because we treat the connections as individual connections. If a connection with a request on it gets dropped the request fails and will have to be retried on the next sync pass. If the connection with the index handlers drop we promote another one, send a ClusterConfig there with our current state, and get new IndexUpdates from that point onwards (regardless of what might already have been sent on the dropped connection).

As an additional bonus we can, in principle, have a mix of QUIC and TCP connections (this works today) although the use case is ... unclear :D

calmh commented:
Note that my multiple-QUIC-streams PR which is also somewhere here takes the opposite approach, which is more natural since multiple QUIC streams share one "connection" and can be established at any time from either direction. That was in fact more complicated in some cases, and doesn't apply cleanly to TCP at all. (Though I had hoped it would when I wrote it, that was sort of phase one and this was phase two; two separate attempts at solving a specific use case for a sponsor.)

calmh commented:
I have added a field to the Hello message that indicates the number of connections configured by the peer. This allows a simple form of negotiation:

- If the other side doesn't support multiple connections (implicitly sends `0`), only ever use one.
- If the other side doesn't want multiple connections (sends `1`), only use one.
- If we both support multiple connections and send a value > 1, use the highest -- so that it's possible to increase the number by just changing one side.

Given this, we can default to something like three (one control connection and two data connections) to get some gains in common situations, plus some feature exercise, without breaking connections with older devices.

(There are some fairly large speedups for high latency connections by using more of them.)

imsodin reviewed (commented):
Unfortunately I am still not through reviewing - sending a bunch of comments anyway in case you already want to have a look. However overall comment/review quality might still improve when I actually progress through all of it.

Generally looks quite good: While clearly complex not as complex as I expected nor standing out much compared to the existing model code (except for being (more) unknown) xD

imsodin commented on lib/model/model.go:
```diff
@@ -728,29 +748,61 @@ func (m *model) ConnectionStats() map[string]interface{} {
+			conn := m.connections[connIDs[0]]
+
+			cs.Primary.Type = conn.Type()
+			cs.Primary.IsLocal = conn.IsLocal()
+			cs.Primary.Crypto = conn.Crypto()
+			cs.Primary.Statistics = conn.Statistics()
 			if addr := conn.RemoteAddr(); addr != nil {
-				ci.Address = addr.String()
+				cs.Primary.Address = addr.String()
+			}
```
Nil check seems unnecessary (we also don't do it for secondaries).

And maybe worth adding `newConnectionInfo(protocol.Connection)` to use here and for secondaries.

calmh replied:
I think this came from connections that don't have an address, e.g., `fakeConn`, causing a panic. I'll check it out and fix it on the other side in that case.

imsodin commented on lib/model/model.go:
```diff
@@ -1291,6 +1343,56 @@ func (m *model) ClusterConfig(conn protocol.Connection, cm protocol.ClusterConfi
+	deviceID := conn.DeviceID()
+	connID := conn.ConnectionID()
+
+	m.pmut.RLock()
+	defer m.pmut.RUnlock()
```
Similar concern as in https://github.com/syncthing/syncthing/pull/8918#discussion_r1285198517: I fear that moving the lock inside a helper function increases the chance of this being missed in future changes and thus potentially introducing lock issues.

calmh replied:
(fixd)

imsodin commented on lib/model/model.go:
```diff
@@ -2306,26 +2407,98 @@ func (m *model) AddConnection(conn protocol.Connection, hello protocol.Hello) {
+			}
+		}
+
+		// Close connections with a worse connection priority than best.
```
This feels like something the connection service can and should handle when adding new connections. Couldn't we expose all connections on a `Model.Connections` method which then also removes the need for the new `deviceConnectionCounter` in `connections.service`?

calmh replied:
I think having the model track connections (for the purpose of external consumers) was a misdesign and I was happy to remove it. (Tell, don't ask, etc.) That said, possibly the connection service could do this. I will look into it.

calmh replied:
Still thinking about this one. Certainly, the connection service could track all connections and "reap" ones with worse priority after a while, thus triggering a promotion of some other connection when the model reacts to the close...

imsodin reviewed (commented):
Another small step.

calmh commented on lib/connections/structs.go:
```diff
@@ -88,15 +92,29 @@ func (t connType) Transport() string {
 
+// newConnectionID generates a connection ID. The connection ID is designed
+// to be 1) unique for each connection (even those reusing the same socket
+// address on both sides), 2) sortable so that the connection with the
+// lowest ID will be the primary one. This also coincides with being the
+// oldest connection.
```
In principle there's no strict requirement that the same connection is considered primary on both sides (but it's better, so we avoid delaying index messages with request/responses), we just need to ensure that a given side considers *one* connection primary so that index updates are properly serialized. In practice, we assume that index updates will come in on the primary connection, but if we get index updates somewhere else we move the index handler there (in ensureIndexHandler) so things work out over time. (We should probably mark that connection as promoted then as well, which I see I don't, I'll look into that.)

calmh replied:
So, this has changed, and the way it works now is that both sides send a timestamp, and then each side builds a connection ID based on the sum of the timestamps. This way, both sides gets the same number and will sort the connections the same. It's also somewhat resilient to one side fucking it up (for example always sending `42` as the timestamp), although if the other side is malicious (send `-time.Now().UnixNano()` for example) things will end up less sorted. That said it should work mostly anyway as each side tacks on a random part to avoid the ID being entirely controllable by the other side. (The one thing we need to avoid is collisions, we don't want two connections with the same ID.)

calmh commented on lib/model/model.go:
```diff
@@ -159,9 +161,11 @@ type model struct {
-	conn                map[protocol.DeviceID]protocol.Connection
+	connections         map[string]protocol.Connection // connection ID -> connection
+	deviceConnIDs       map[protocol.DeviceID][]string // device -> connection IDs (invariant: if the key exists, the value is len >= 1, with the primary connection at the start of the slice)
+	promotedConnID      map[protocol.DeviceID]string   // device -> latest promoted connection ID
```
Might also be something other than the first element in deviceConnIds if that connection has been closed but we haven't run promotion yet. I think the explanation here gets a little long-winded for one line, perhaps it's better to keep the commentary in the promotion method?

calmh commented on lib/model/model.go:
```diff
@@ -2984,6 +3184,18 @@ func (m *model) CommitConfiguration(from, to config.Configuration) bool {
 	return true
 }
 
+func (m *model) setConnRequestLimiters(cfg config.DeviceConfiguration) {
+	m.pmut.Lock()
```
Agreed, I'll look into it.

calmh replied:
(resolved)

calmh commented on lib/model/model.go:
```diff
@@ -2306,26 +2407,98 @@ func (m *model) AddConnection(conn protocol.Connection, hello protocol.Hello) {
+}
+
+func (m *model) scheduleConnectionPromotion() {
+	m.promotionTimer.Reset(time.Second)
```
Well... It's a one second schedule after adding or removing connections. The reason for doing this is that if a device shuts down and we have 16 connections to it, we will get 16 connection closes in rapid succession. We don't want/need to run the promotion process 16 times during that time, but we do want to run it once things have quieted down. I think if the connection frequency is such that "after one second" becomes "never" then there's a different problem to be solved...

As for AfterFunc vs in serve, I think it's nice to have it called in one place under known circumstances, instead of spread out / scheduled in places where it's not visible. Plus we'd need to track the timer returned by AfterFunc to be able to cancel it and renew, which would need locking and whatnot...

imsodin replied:
Oh I totally didn't get that it's a feature that it keeps deferring the execution, not an unfortunate side-effect.  That definitely makes the `AfterFunc` point entirely moot. Eternal delay... Yeah I guess we can ignore that, even with tons of devices that really shouldn't be a thing (also some bad connection cycle should still be >>1s).

I'll give it another shot at proposing a comment, lets see if this turns out comprehensible (and without missed gotchas) :) 

```suggestion
	// Keeps deferring to prevent multiple executions in quick succession, e.g. if multiple connections to a single device are closed.
	m.promotionTimer.Reset(time.Second)
```

calmh commented on lib/model/model.go:
```diff
@@ -1777,32 +1879,54 @@ func (m *model) introduceDevice(device protocol.Device, introducerCfg config.Dev
+	remainingConns := without(m.deviceConnIDs[deviceID], connID)
+	if removedIsPrimary {
+		m.progressEmitter.temporaryIndexUnsubscribe(conn)
+		if idxh, ok := m.indexHandlers[deviceID]; ok && idxh.conn.ConnectionID() == connID {
+			delete(m.indexHandlers, deviceID)
+		}
```
This might be for the case where we've moved the index handler to a different, supposedly non-promoted, connection due to getting index updates there instead... If so, it couldn't hurt with some commentary.

Yes, I was also a little bit confused about the index handler lifetime.

imsodin replied:
> This might be for the case where we've moved the index handler to a different, supposedly non-promoted, connection due to getting index updates there instead... If so, it couldn't hurt with some commentary.

Definitely need to look into that (moving index handler), I managed to not even notice that behaviour/path while reviewing until you mentioned it in a comment above and here.

imsodin replied:
Oh also the index handler stuff is the (or at last a) cause for the flaky cluster config test that came up somewhere else - I'll PR a fix for that, the conflicts won't be bad.

calmh replied:
I think this is fine now. One thing I may want to do in the long run, but which didn't look trivial right now, is to separate the sending of indexes with the receiving. These are in principle two separate processes that don't need to be intertwined, but currently they are both bundled in the indexhandler. If they were separated then it wouldn't matter at all if both sides agreed on a "primary" connection, other than for the slight latency advantage it might have to avoid blocking the index-sending connection with data-sending.

calmh commented on lib/model/model.go:
```diff
@@ -1777,32 +1879,54 @@ func (m *model) introduceDevice(device protocol.Device, introducerCfg config.Dev
+		// All device connections closed
+		delete(m.deviceConnIDs, deviceID)
+		delete(m.promotedConnID, deviceID)
+		delete(m.connRequestLimiters, deviceID)
```
Maybe. Let me think on this a bit.

calmh commented on lib/model/model.go:
```diff
@@ -1777,32 +1879,54 @@ func (m *model) introduceDevice(device protocol.Device, introducerCfg config.Dev
 
-	m.progressEmitter.temporaryIndexUnsubscribe(conn)
-	m.deviceDidClose(device, time.Since(conn.EstablishedAt()))
+	k := map[bool]string{false: "secondary", true: "primary"}[removedIsPrimary]
```
Sorry

calmh commented:
Lots of stuff to consider, which I will do and ask for a new look. About the locking stuff though, I wonder if that's something that we should spend some engineering on improving separately. I'll write some thoughts here though since we're talking about it.

- We have two locks, fmut (folder related stuff) and pmut (connection related stuff, who knows what the `p` might have meant originally)
- Sometimes we need both, and for deadlock avoidance it's critical to always lock in the same order, which we've standardized on being alphabetical -- so it's OK to grab `pmut` while holding `fmut` but never the opposite.
- So, currently, it's always safe to grab `pmut` without knowing the context, e.g., from small methods like the ones under review here. Conversely, care must be taken when grabbing `fmut`...
- Many of the locks are just to protect map accesses, e.g. `m.fmut.Rlock(); runner, ok := m.folderRunners[folder]; m.fmut.RUnlock()`.

I wonder if we could get away with a single, bigger lock used for when consistency is required (we're going to change a lot of internal state and need to be alone while doing it) and switching to atomic maps for most other small accesses. Those can be easily type safe these days.

Or, potentially, making the model "single threaded" with some serve loop instead of all the things running concurrently, though that's bigger change.

And none of this specific to this PR anyway. :)

calmh commented:
I added a timestamp field to the Hello message, which allows both sides to agree on a value to use for the connection ID. Thus the potential mismatch of which connection is "primary" goes away, and the reasoning becomes easier in the rest of the code. That said, I don't think there is any value in removing the handling that is present to be agile when the primary connection changes.

imsodin commented:
The hello refactor is a nice cleanup/simplification!

Given the suspicious behaviour debugging and a number of open review comments, I assume it's still wip and you'll ping again for another round (if I don't react to updates to the PR otherwise).

calmh commented:
Yeah I'm going to merge and address stuff this afternoon and then go for a litter round

calmh commented on lib/connections/service.go:
```diff
@@ -666,8 +669,17 @@ func (s *service) resolveDialTargets(ctx context.Context, now time.Time, cfg con
+			continue
+		}
+		if currentConns > 0 && !dialer.AllowsMultiConns() {
+			l.Debugf("Not dialing using %s as it does not allow multiple connections and we already have %d/%d connections", dialerFactory, currentConns, deviceCfg.MultipleConnections)
```
(resolved)

calmh commented:
@imsodin 👋

calmh commented:
@imsodin I removed the connection reaping from the model, instead having the connection service do that. Now it closes bad connections when adding a new better one.

imsodin commented on lib/sliceutil/sliceutil.go:
```diff
@@ -0,0 +1,16 @@
+
+package sliceutil
+
+// RemoveAndZero removes the element at index i from slice s and returns the
+// resulting slice. The slice ordering is preserved; the last slice element
+// is zeroed before shrinking.
+func RemoveAndZero[E any, S ~[]E](s S, i int) S {
+	copy(s[i:], s[i+1:])
+	s[len(s)-1] = *new(E)
+	return s[:len(s)-1]
+}
```
I am missing something about the zero-ing: We are zeroing the element at the end, but that element is still referenced at the second to last (resp. new last in the truncated slice) - as in it can't be GCed. Why are we zeroing then?

calmh replied:
This is the typical slice removal pattern. :) It's just about making sure elements no longer part of the slice (but still present in the backing array) do not contain pointers pointing to things that prevent them from being GC:d.

imsodin replied:
I get the general mechanics, it's just the zeroing that confused me - I think I get it now. I think the zeroing only has an effect though if we are removing the last element - otherwise the element in the backing array after the slice end is also present in the slice due to the copy - so zeroing it out doesn't do anything for GC. Only if it happens to be the last element that element would persist in the array but not the slice. Hardly worth "optimising" (moving the zeroing into a `if i == len(s) - 1` block) though, I am just happy I didn't have a fundamental misunderstanding here (hopefully) :)

calmh replied:
> otherwise the element in the backing array after the slice end is also present in the slice due to the copy - so zeroing it out doesn't do anything for GC

Not this cycle, no, but then you've set it up for *that* element never being GC:d.

calmh commented:
Anyhow, what about the rest? Are we good here? Are you looking for me to fix something?

imsodin commented:
Just looking for me to have a proper look at the (good looking and extensive) changes since my last review.

imsodin commented on lib/connections/service.go:
```diff
@@ -545,19 +581,22 @@ func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Con
 			priorityCutoff -= cfg.Options.ConnectionPriorityUpgradeThreshold
 
-			if bestDialerPriority >= priorityCutoff {
+			currentConns := s.numConnectionsForDevice(deviceCfg.DeviceID)
```
Nit: Could define that on L584 already.

imsodin commented on lib/connections/service.go:
```diff
@@ -669,8 +708,17 @@ func (s *service) resolveDialTargets(ctx context.Context, now time.Time, cfg con
 		priority := dialer.Priority(uri.Host)
-		if priority >= priorityCutoff {
-			l.Debugf("Not dialing using %s as priority is not better than current connection (%d >= %d)", dialerFactory, priority, priorityCutoff)
+		currentConns := s.numConnectionsForDevice(deviceCfg.DeviceID)
+		if priority >= priorityCutoff && currentConns >= s.desiredConnectionsToDevice(deviceCfg.DeviceID) {
+			l.Debugf("Not dialing %s at %s using %s as priority is not better than current connection (%d >= %d) and we already have %d/%d connections", deviceID.Short(), addr, dialerFactory, priority, priorityCutoff, currentConns, deviceCfg.NumConnections)
+			continue
+		}
+		if priority > priorityCutoff {
+			l.Debugf("Not dialing %s at %s using %s as priority is worse than current connection (%d > %d)", deviceID.Short(), addr, dialerFactory, priority, priorityCutoff)
+			continue
+		}
```
The order with overlapping priority conditions had me confused for a second - might be nicer to change the order of conditions and use `==` then.

calmh replied:
Clarified!

imsodin commented on lib/model/model.go:
```diff
@@ -1138,17 +1184,12 @@ func (m *model) handleIndex(conn protocol.Connection, folder string, fs []protoc
-		m.evLogger.Log(events.Failure, "index sender does not exist for connection on which indexes were received")
-		l.Debugf("%v for folder (ID %q) sent from device %q: missing index handler", op, folder, deviceID)
-		return fmt.Errorf("index handler missing: %s", folder)
+		l.Infof("%v for folder %s sent from %s (%s), but no index handler is registered for this connection.", op, folder, deviceID.Short(), conn)
+		return fmt.Errorf("%s: %w", folder, ErrFolderNotRunning)
```
Isn't this still a failure, as in should never happen? The remote should only send us an index after successful cluster config exchange, and that only happens during connection promotion where the index handler is also added.

calmh replied:
I believe so; I don't remember why I changed commentary etc, reverting.

imsodin reviewed (commented):
Looks good (and scary) - few small comments.

There's again/still a few not directly related refactors (slice removal, short device IDs, var declarations in `handleHellos`, ...) - only if you feel like it, those would be nice to have committed separately.

calmh commented on lib/connections/service.go:
```diff
@@ -315,17 +346,19 @@ func (s *service) connectionCheckEarly(remoteID protocol.DeviceID, c internalCon
+			l.Debugf("Not accepting connection to %s at %s: already have %d connections, desire %d", remoteID, c, currentConns, desired)
 			return errDeviceAlreadyConnected
 		}
 	}
```
I refactored it to be a single check, at the price of calculating more stuff up front. These are not expensive things though, and we only do it once per connection, so I think it's alright.

calmh commented:
> There's again/still a few not directly related refactors (slice removal, short device IDs, var declarations in handleHellos, ...) - only if you feel like it, those would be nice to have committed separately.

Frankly it would be a huge pain in the ass to extract as it's all interleaved, and the formatting comes from updated editor settings, so I would *very* much prefer to keep it in... 🙏

imsodin reviewed (commented):
lgtm except that I now also looked at the UI code - only one little confusion of mine there.

Are you planning to wait with merging this until after rc2? Feels like this is a scary enough change to get it released at the earliest possible time, i.e. rc1 of the next release cycle.

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
+                  <input name="numConnections" id="numConnections" class="form-control" type="number" pattern="\d+" ng-model="currentDevice.numConnections" min="1" />
+                </div>
+              </div>
+              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a positive number</p>
```
```suggestion
              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a positive number.</p>
```

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
+                </div>
+              </div>
+              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a positive number</p>
+              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine">When set to more than one, Syncthing will attempt to establish multiple concurrent connections. One connection is reserved for control messages, the rest are used for data transfer.</p>
```
What exactly are "control messages"? I've got a feeling the user won't be able to understand part without a longer explanation (e.g. a through a link to the Docs).

calmh replied:
I don't think it matters too much if the precise definition is understood as it's followed by "the rest are used for data transfer", but feel free to suggest something. As for when it's set to one, yes, obviously everything will run on that one connection like it's always done.

tomasz1986 replied:
I would just add this clarification, but when it comes to "control messages", I can't really suggest anything, as I'm also not sure what exactly they are 🙁.

```suggestion
              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine">When set to more than one, Syncthing will attempt to establish multiple concurrent connections. One connection is reserved for control messages, the rest are used for data transfer. When set to one, the single connection is used for both.</p>
```

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -150,14 +161,15 @@
                     </div>
                   </div>
                   <p class="help-block" ng-if="!deviceEditor.maxSendKbps.$valid && deviceEditor.maxSendKbps.$dirty" translate>The rate limit must be a non-negative number (0: no limit)</p>
+                  <p class="help-block" ng-if="(deviceEditor.maxSendKbps.$valid || deviceEditor.maxSendKbps.$pristine) && (deviceEditor.maxRecvKbps.$valid || deviceEditor.maxRecvKbps.$pristine)">The rate limit is applied to the aggregate of all connections to this device.</p>
```
Is there a simpler word than "aggregate"? The non-technical user likely won't have a clue what this means.

calmh replied:
Suggestion?

tomasz1986 commented on gui/default/index.html:
```diff
@@ -871,6 +871,14 @@ <h4 class="panel-title">
+                      <tr ng-if="connections[deviceCfg.deviceID].connected && (deviceCfg.numConnections > 1 || connections[deviceCfg.deviceID].secondary)">
+                        <th><span class="fas fa-fw fa-random"></span>&nbsp;<span translate>Number of Connections</span></th>
+                        <td class="text-right">
+                          <span>
```
I don't think the empty `span` is needed here.

calmh replied:
Improved

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
+              <div class="row">
+                <span class="col-md-8" translate>Number of Connections</span>
+                <div class="col-md-4">
+                  <input name="numConnections" id="numConnections" class="form-control" type="number" pattern="\d+" ng-model="currentDevice.numConnections" min="1" />
```
Not really familiar with the rest of the code, but maybe allow the value of `0` to imply the default (i.e. `3` at the moment)? This is currently the case with many other settings (although not all of them, to be exact).

calmh replied:
The default gets baked when loading so that wouldn't help in this case. I think our handling of defaults is inconsistent and bad and it's on my list for future improvements, but it's tricky...

bt90 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -150,14 +161,15 @@
                     </div>
                   </div>
                   <p class="help-block" ng-if="!deviceEditor.maxSendKbps.$valid && deviceEditor.maxSendKbps.$dirty" translate>The rate limit must be a non-negative number (0: no limit)</p>
+                  <p class="help-block" ng-if="(deviceEditor.maxSendKbps.$valid || deviceEditor.maxSendKbps.$pristine) && (deviceEditor.maxRecvKbps.$valid || deviceEditor.maxRecvKbps.$pristine)">The rate limit is applied to the aggregate of all connections to this device.</p>
```
```suggestion
                  <p class="help-block" ng-if="(deviceEditor.maxSendKbps.$valid || deviceEditor.maxSendKbps.$pristine) && (deviceEditor.maxRecvKbps.$valid || deviceEditor.maxRecvKbps.$pristine)">The rate limit is applied to the accumulated traffic of all connections to this device.</p>
```

calmh commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
-            <div class="col-md-12">
+          <div class="row">
+            <div class="col-md-6" ng-class="{'has-error': deviceEditor.numConnections.$invalid && deviceEditor.numConnections.$dirty}">
+              <label translate>Connection Management</label>
```
Not having a header makes the layout odd IMHO, what's your suggestion?

<img width="923" alt="Screenshot 2023-09-04 at 11 28 58" src="https://github.com/syncthing/syncthing/assets/125426/a8e6d1c7-f1eb-49ce-a34e-47f9c282b62f">

tomasz1986 replied:
A different approach, but maybe this change warrants adding a new "Connections" tab where both the number of connections and rate limits would go? Similarly as it's the case in the generic GUI Settings.

tomasz1986 replied:
Yeah, so the decision is about what belongs to "Connections" settings intended to be tweaked by the average user, and "Advanced" settings intended for the power user. Now thinking about it, the "Number of Connections" likely belongs to the second category, as the average user shouldn't really touch it.

(off-topic)
There is a discrepancy regarding rate limits right now, as in the generic settings, they are under "Connections", and here they are under "Advanced", but this is probably something to address in a different PR.

calmh commented:
> Are you planning to wait with merging this until after rc2? Feels like this is a scary enough change to get it released at the earliest possible time, i.e. rc1 of the next release cycle.

@imsodin fully agreed. There is/was no rc.2 this cycle because the timing was already wrong due to last month's delays. The current rc.1 goes out as the release tomorrow. I would like to see this merged early for 1.25.0, next cycle.

calmh commented:
> With pure metadata connections being a thing, should we reconsider enabling TCP nodelay?

@bt90 consider away, my gut feeling says it has no effect and implementation would be tricky, but I'm open to being proven wrong. :)

bt90 commented:
Toggling it for all connections is easy:

https://github.com/syncthing/syncthing/blob/4138e22898c4b226a029dc6cc435496857b74def/lib/dialer/public.go#L35

My last tests showed no difference for a single connection so I doubt that this has a negative impact on the rather bulk-ish data connections. The question is if the "anemic" control connections would benefit from it.

calmh commented:
I don't think they would. The advantage I can see with separating control/metadata messages (essentially, index updates) from request/response messages is that the queue for response messages can get quite large time wise: several minutes, when the connection is slowish and there is a lot of data being transferred. Keeping it separates propagates those updates quicker.

The delay/nodelay stuff is on the order of hundreds of milliseconds though? This will pretty much never matter, but sending our messages as several small writes might be a disadvantage. In any case, separate from this PR.

calmh commented:
I changed the handling of the default value. Now it remains zero in the config and Syncthing interprets this how it wants. I changed it back to be interpreted as one. This means the multiple connection stuff will get less exercise in the RC, but also that people opposed to it will not get a rude awakening, and any errors in the mechanism itself will be easier to spot, perhaps.

I still think >1 is a better default in the future, but we can wait a few releases to do that -- similar to how we handled large blocks, etc.

bt90 commented:
> I changed it back to be interpreted as one. This means the multiple connection stuff will get less exercise in the RC, but also that people opposed to it will not get a rude awakening, and any errors in the mechanism itself will be easier to spot, perhaps.

My feelings are somewhat mixed. I agree wholeheartedly that >1 is the better default and can see why we might want to play it safe here.

Will the default remain as it is, or would clients end up with the fallback value of `1` written into their config files?

calmh commented:
@bt90 we passed each other because I waffled and deleted the comment you replied to... But no, we would now *not* bake in the `1`, so it would remain `0` and be open to interpretation.

bt90 commented:
Sounds good :+1:

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
+                  <input name="numConnections" id="numConnections" class="form-control" type="number" pattern="\d+" ng-model="currentDevice.numConnections" min="0" />
+                </div>
+              </div>
+              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a positive number.</p>
```
```suggestion
              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a non-negative number.</p>
```

tomasz1986 commented on gui/default/syncthing/device/editDeviceModalView.html:
```diff
@@ -137,11 +137,22 @@
+                </div>
+              </div>
+              <p class="help-block" ng-if="!deviceEditor.numConnections.$valid && deviceEditor.numConnections.$dirty" translate>The number of connections must be a non-negative number.</p>
+              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine" translate>When set to more than one, Syncthing will attempt to establish multiple concurrent connections. One connection is reserved for control messages, the rest are used for data transfer. When set to one, the single connection is used for both. Set to zero to let Syncthing d ...
```
```suggestion
              <div class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine">
                <p translate>When set to more than one, Syncthing will attempt to establish multiple concurrent connections. For this to succeed, the number must be set to the same value on both devices.</p>
                <p translate>One connection is reserved for control messages, the rest are used for data transfer. When set to one, the single connection is used for both. Set to zero to let Syncthing decide.</p>
              </div>
```

I've added "For this to succeed, the number must be set to the same value on both devices." as I think this is crucial information here, and I've also split the text into two paragraphs, as it seemed a bit too long (and problematic when translating when changing a single word would force to re-translate the whole thing).

calmh replied:
It doesn't have to be the same value, though. I think this is already too long to be the description text here. How about walking back a lot of this to just something like `How many parallel connections to use with this device, or zero to let Syncthing decide.` and the usual help icon that points to the documentation article?

tomasz1986 replied:
Honestly, I'm not a fan of moving everything to the Docs, because a) the Docs tend to be much more technical, and b) the Docs are English-only.

How about a shorter version, e.g. like this:

```suggestion
              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine" translate>When set to more than one, Syncthing will attempt to establish multiple concurrent connections. For this to work, the number must be at least equal on both devices. Set to zero to let Syncthing decide.</p>
```

calmh replied:
It doesn't *have to* be at least equal either, it'll work anyway just with fewer connections and now it sounds threatening :D

Sorry, I realise it's really tricky to be concise and yet convey the desirable meaning. Im not sure what to say here.

tomasz1986 replied:
Understood 😉. Yet another try:

```suggestion
              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine" translate>When set to more than one, Syncthing will attempt to establish multiple concurrent connections. If not equal on both devices, the lower value will be used. Set to zero to let Syncthing decide.</p>
```

calmh replied:
It's actually the higher value, if both sides are set to a value higher than one, otherwise one... 🥲 But... I wouldn't be surprised if this is one of those things we change when this gets deployed...

tomasz1986 replied:
Is it like that? I mean, if one device is set to `2` and the other to `3`, what happens then?

calmh replied:
Three

tomasz1986 replied:
I see… Is it safe like this? For example, if I'm connected to someone else's device (e.g. a friend), and they set the number of connections to a crazy value like `1000000` (let's say just by mistake), will this essentially break Syncthing on my side?

calmh replied:
Probably! Maybe we should have an upper limit.

tomasz1986 replied:
```suggestion
              <p class="help-block" ng-if="deviceEditor.numConnections.$valid || deviceEditor.numConnections.$pristine" translate>When set to more than one on both devices, Syncthing will attempt to establish multiple concurrent connections. If the values differ, the highest will be used. Set to zero to let Syncthing decide.</p>
```

calmh replied:
I added a limit below.

imsodin reviewed (approved):
Sorry, I thought I already approved o.O

brummbrum commented:
THANK YOU, THANK YOU! I only tried this today and it makes a world of difference with my ISP. Finally, I can saturate my contracted upload bandwidth over the internet. I have a 100Mb/s down & 10Mb/s up contract on two different sites. Unfortunately my ISP (deliberately or accidentally) caps uploads at approx 2Mb/s per connection. Download rates are no problem, I can saturate 100Mb/s with just a single link. But single connection uploads suck.
I have found that this is a quite common problem for many users with different ISPs. It seems that from an ISP perspective they mostly deal with folks that are really only interested in download speeds. The few geeks who are self-hosters or whatever are simply getting under the bus. This makes it cheaper for ISPs because they need less expensive peering agreements with other network operators.
Now that I can set numConnections between the different sites (syncthing devices) my synchronization speeds are finally at the maximum possible 10Mb/s speed over the WAN!! I know this took a long time and I saw some heated forum discussions related to this topic. But finally, it is working. Thanks again!
